src/watcher.rs

```diff
@@ -0,0 +1,319 @@
+                self.watcher.watch(dirname, RecursiveMode::NonRecursive)?;
+
+                let mut files = HashSet::new();
+                files.insert(name.to_string());
+                dirs.insert(
+                    dirname.to_owned(),
+                    DirWatch {
+                        full_directory: false,
+                        files,
+                    },
+                );
+                Ok(())
+            }
+        }
+    }
+
+    /// 'path' must be a directory, not a file.
+    pub fn watch_dir<P: AsRef<Path> + std::fmt::Debug>(
+        &mut self,
+        path: P,
+        recurse: RecursiveMode,
+    ) -> Fallible<()> {
+        self.unwatch_info
+            .insert(path.as_ref().to_owned(), UnwatchInfo::Directory(recurse));
+
+        match recurse {
+            RecursiveMode::Recursive => match &mut self.recursive_watcher {
+                Some(w) => {
+                    w.watch(path, RecursiveMode::Recursive)?;
+                    Ok(())
+                }
+                None => {
+                    let mut w = notify::watcher(self.event_sink.clone(), self.delay)?;
+                    w.watch(path, RecursiveMode::Recursive)?;
+                    self.recursive_watcher = Some(w);
+                    Ok(())
+                }
+            },
+
+            RecursiveMode::NonRecursive => {
+                let mut dirs = self.dirs.lock().expect("FSWatch: mutex error");
+                match dirs.get_mut(path.as_ref()) {
+                    Some(dw) => {
+                        dw.full_directory = true;
+                        Ok(())
+                    }
+                    None => {
+                        // watch first; if this throws an error, don't insert into the 'dirs' structure
+                        self.watcher.watch(&path, RecursiveMode::NonRecursive)?;
+
+                        dirs.insert(
+                            path.as_ref().to_owned(),
+                            DirWatch {
+                                full_directory: true,
+                                files: HashSet::new(),
+                            },
+                        );
+                        Ok(())
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn unwatch<P: AsRef<Path>>(&mut self, path: P) -> Fallible<()> {
+        let info = match self.unwatch_info.get(path.as_ref()) {
+            Some(info) => info,
+            None => return Err(failure::err_msg("FSWatch::unwatch on a non-watched path")),
+        };
+
+        let (key, filename): (&Path, Option<&str>) = match info {
+            UnwatchInfo::File(key, filename) => (&key, Some(filename)),
+            UnwatchInfo::Directory(RecursiveMode::NonRecursive) => (path.as_ref(), None),
+            UnwatchInfo::Directory(RecursiveMode::Recursive) => {
+                self.recursive_watcher
+                    .as_mut()
+                    .expect("FSWatch: inconsistent UnwatchInfo")
```

Comment:
I honestly don't understand what the possible consequences of not shutting down the server would be in this case but my first thought would be to handle this case and return an error instead of `expect`ing it. The `expect` calls on the `lock` calls I think are ok, but this one I'm not so sure.

Other than that this looks good to me, although @autozimu might be a better judge for this one.

Reply:
For both of the "inconsistent UnwatchInfo" `expect` calls, hitting them would mean that one the one hand we think there should be a watcher set up for the `unwatch`ed path, but on the other hand we find no such watcher in our data structures. This would mean a bug in the (my) watcher code.

If we don't panic there but just return an error, we are apparently in a situation where we're not sure exactly what watchers are or aren't active on what paths. In the end, I believe this would mean that the language server may get file change events that it didn't ask for, or may _not_ get events for paths that it does have watched.

I'm not sure what your policy is in such situations. It might be possible to merge the two parallel data structures (`dirs` and `unwatch_info`), so that this case cannot occur. However, I'm personally not sure exactly how that would play with the `Path` interface; e.g. do `parent` and `file_name` behave predictably and sanely for multiple different ways of writing the same path? That is ultimately why I opted for the `unwatch_info` strategy.

But by all means, if you disagree this can change :)

Reply:
Firstly, thank you for taking time to tackle this.

On the point of `expect` calls, I would prefer to reduce the number of "hard errors" as much as possible. I do agree that in the unexpected case, we can only issuing an error and nothing more. But halting the whole program is a worse user experience, which also hinder future sophisticated error handling such as restarting the watch thread.

I haven't look through the the whole changes yet and I don't know much further on the `parent` and `file_name` topic. Will look further and update here.

Reply:
@autozimu Sorry to poll here, but were you intending to do some research and post here again, or are you waiting for me to do anything? I've been happily using my branch for a while now, and it seems to be working (though that doesn't mean the code is perfect, just a confirmation that it's not total crap).